twister: Implement GNU make jobserver support

jeremybettis opened the pull request:
When twister is started from make, use the information in MAKEFLAGS to participate in the job server started by make.

When twister is running standalone, start a jobserver to share with child processes. If using the kitware version of ninja, then this will prevent ninja+twister from starting more processes than XX when twister is run with -jXX.

There is no implementation for windows yet.

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis committed:
twister: Implement GNU make jobserver support

When twister is started from make, use the information in MAKEFLAGS to
participate in the job server started by make.

When twister is running standalone, start a jobserver to share with
child processes. If using the kitware version of ninja, then this will
prevent ninja+twister from starting more processes than XX when twister
is run with -jXX.

There is no implementation for windows yet.

Signed-off-by: Jeremy Bettis <jbettis@google.com>

jeremybettis commented:
I tested with this:

```
twister -v -i --clobber -j1 &> /dev/null &
while jobs %% ; do uptime; jobs -x pstree -T -c %% | cat -n ; sleep 1 ; done
```

And verified that twister+cmake+ninja only ran 10 processes. There were 11 for most of the time, but I think that the extra one is twister's main thread that is just waiting at line 1029, so that still seems fine.

I'm using cmake 3.20.5 and ninja 1.10.2.git.kitware.jobserver-1

If there are additional scenarios you would like to see please reply to this comment.

hakehuang commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -185,7 +188,7 @@ class CMake:
     dt_re = re.compile('([A-Za-z0-9_]+)[=]\"?([^\"]*)\"?$')
 
-    def __init__(self, testsuite, platform, source_dir, build_dir):
+    def __init__(self, testsuite, platform, source_dir, build_dir, jobserver):
```
can we make this jobserver as a decoration? like
```
@jobserver
def parse_generated(self):
```

andyross reviewed (dismissed):
This seems to be working great for me, so +1, though the python people will definitely want to go over it; this isn't my world.

The requirement for a forked ninja binary (for those who have trouble finding it: clone https://github.com/Kitware/ninja and build from source -- I didn't find a binary download link a-la their cmake distro) is really unfortunate though.  A default twister run on a big host remains pretty broken.

Can the "jobserver ninja" variant be detected dynamically via --version or something?  We could have twister check for it and degrade to "-j 1" if it is stuck with "relentlessly parallel ninja".

Also: ninja is very small; it would strike me as very feasible to add this version to the SDK.  @stephanosio?

andyross commented:
Also, and this isn't really anything to do with this patch but it's part of the same problem: having a jobserver for build jobs is great, but it doesn't capture everything.  In fact twisters own threads (there are a ton) consume significant CPU while parsing the emulator output, and the SMP emulators are "one job" as far as make/ninja is concerned, but actually spawn one thread per CPU internally.  There's no attempt made to account for this in twister, so the load routinely runs just a tiny bit higher than we want it (e.g. ~136 on my device with 128 host CPUs).  This doesn't sound like a big problem, but the moment you go beyond the hardware limits threads start getting scheduled out by the host kernel, which means we start seeing spurious SMP failures due to races between the two CPU threads.

It would be nice to have some kind of accounting for this; we could maybe move from threads to asyncio (which is a bunch of boilerplate to add "async/await" keywords and to swap the read() calls for their async equivalents, but in practice doesn't change the structure of the code), which would put all of twister in "1 job" which we could account for.  And then we could just have some logic that knows "make run on qemu SMP needs extra jobs" and do the accounting internally).

jeremybettis commented:
> Also: ninja is very small; it would strike me as very feasible to add this version to the SDK. @stephanosio?

pip install ninja will give you  `1.10.2.git.kitware.jobserver-1`

Following the Zephyr getting started guide will give you the stock ninja.  Perhaps we could just update the instructions to use pip install or to download directly from kitware's releases at https://github.com/Kitware/ninja/releases instead?

keith-zephyr commented on scripts/pylib/twister/twisterlib/jobserver.py:
```diff
@@ -0,0 +1,223 @@
+# Copyright 2022 The ChromiumOS Authors
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+# SPDX-License-Identifier: Apache-2.0
```
This needs to be cleaned up.  Zephyr's LICENSE file is specifically Apache-2.0, while the [Chromium LICENSE](https://source.chromium.org/chromiumos/chromiumos/codesearch/+/HEAD:src/platform/ec/LICENSE) file is the BSD-style license.
This probably should just be `SPDX-License-Identifier: BSD-3-Clause`.

keith-zephyr replied:
Embedding the Chromium LICENSE directly might be appropriate, similar to [strtoul.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtoul.c).  But I'm not a lawyer.

jeremybettis replied:
done

jeremybettis replied:
Google lawyers said to use 

```
# Copyright 2022 Google LLC.
# SPDX-License-Identifier: Apache-2.0
```

So that is better for Zephyr also.

teburd commented:
The python code looks ok to me, though I'm not a python person by any means either... its a bummer that stock ninja doesn't respect the job server.

jeremybettis commented:
I fixed a bug when running twister from a Makefile without the + sign, now it checks if the passed file descriptors are readable and writable as expected, rather than just checking if the file descriptors exist.

Then I rebased my branch to head to deal with the merge conflict, and somehow the review bot added 90 reviewers, sorry.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -4,30 +4,33 @@
-import sys
-import subprocess
 import pickle
-import logging
```
please, when do this type of change (sorting), do that in a separate commit. It is difficult to see how this is related to the original change and what you are adding/removing here.

jeremybettis replied:
Is that a requested change, or just some advice?

nashif replied:
advice :)

nashif replied:
still in the process of testing this in a production env.

fabiobaltieri replied:
Any feedback? This has been sitting for a while.

laurenmurphyx64 replied:
Pinging @nashif for assignee approval before feature freeze (sorry if this is annoying)

nordicjm commented:
Gave this a try on twister with system ninja (arch linux, which has support for -j argument) and ran some tests limiting it to 3 jobs, twister confirmed it was running 3 jobs, however system load quickly went to 20 (this is an 8 core system with hyper threading, would expect to see a system load of 6 or 12 if the job limiting was working correctly) so it does not look like this works for me.

teburd commented:
> Gave this a try on twister with system ninja (arch linux, which has support for -j argument) and ran some tests limiting it to 3 jobs, twister confirmed it was running 3 jobs, however system load quickly went to 20 (this is an 8 core system with hyper threading, would expect to see a system load of 6 or 12 if the job limiting was working correctly) so it does not look like this works for me.

It sounds like it was working?  make with -j limits the concurrency to -j not the system load.

nordicjm commented:
> It sounds like it was working? make with -j limits the concurrency to -j not the system load.

Concurrency of 3 should spawn 3 threads, each thread 1x CPU core, so system load should not be 20, which indicates equivalent of 20 processes want 100% of a single CPU core.

teburd commented:
> > It sounds like it was working? make with -j limits the concurrency to -j not the system load.
> 
> Concurrency of 3 should spawn 3 threads, each thread 1x CPU core, so system load should not be 20, which indicates equivalent of 20 processes want 100% of a single CPU core.

Did you use the patched ninja that support the job server protocol? Kitware has a fork for this as noted in the pr. Arch Linux ninja package isn’t the forked version (I use arch as well and had to install the fork)

nashif commented:
after many measurements done on a large system (>100 cores), here is a brief summary:

- Using make instead of ninja is the way to go if you want to keep the load under control (make was the default for years)
- But, make is at least 50s slower in the subset I tested (qemu_x86)
- Using the patched version of ninja helps, without this PR
- this PR with `make` has no effect almost, but it works well with default ninja


The last few bumps in the graph below are with make, everythng else is with ninja, with and without this PR and with and without the patched ninja

![image](https://user-images.githubusercontent.com/180017/214731673-3a0f8567-dfae-4b30-9a94-51ba72ac911b.png)

The first spike is what I tried to first with a large job and with all default, cold ccache

nashif commented on scripts/pylib/twister/twisterlib/jobserver.py:
```diff
@@ -0,0 +1,236 @@
+# SPDX-License-Identifier: Apache-2.0
+"""Module for job counters, limiting the amount of concurrent executions."""
+
+import fcntl
```
yep, just noticed that...
```
Traceback:
C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests\test_twister.py:7: in <module>
    from twister_cmd import Twister
twister_cmd.py:22: in <module>
    from twisterlib.twister_main import main
..\pylib\twister\twisterlib\twister_main.py:19: in <module>
    from twisterlib.runner import TwisterRunner
..\pylib\twister\twisterlib\runner.py:25: in <module>
    from twisterlib.jobserver import GNUMakeJobClient, GNUMakeJobServer, JobClient
..\pylib\twister\twisterlib\jobserver.py:5: in <module>
    import fcntl
E   ModuleNotFoundError: No module named 'fcntl'
```

nashif replied:
hmm, looks like we need a windows test for twister to catch this type of issues.

stephanosio replied:
Hotfix: https://github.com/zephyrproject-rtos/zephyr/pull/54170

stephanosio commented:
This PR also seems to have broken macOS support:
https://github.com/zephyrproject-rtos/sdk-ng/actions/runs/4016970387/jobs/6915967155#step:14:130
